analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/SecureRandomSeedsShouldNotBePredictable.cs

```diff
@@ -145,19 +165,40 @@ private static ProgramState ProcessSecureRandomSetSeed(ProgramState state, IInvo
             {
                 constraint = value;
             }
-            return state.SetSymbolConstraint(invocation.Instance.TrackedSymbol(state), constraint);
+            return state.SetSymbolConstraint(instanceSymbol, constraint);
         }
         return null;
+
+        bool IsSetSeed() =>
+            invocation.TargetMethod.Name == "SetSeed"
+            && IsSecureRandom(invocation);
+
+        bool IsAddSeedMaterial() =>
+            invocation.TargetMethod.Name == "AddSeedMaterial"
+            && invocation.Instance is { } instance
+            && instance.Type.Is(KnownType.Org_BouncyCastle_Crypto_Prng_IRandomGenerator);
     }
 
-    private ProgramState ProcessSecureRandomNext(ProgramState state, IInvocationOperationWrapper invocation)
+    private ProgramState ProcessNextMethods(ProgramState state, IInvocationOperationWrapper invocation)
     {
-        if (SecureRandomNextMethods.Contains(invocation.TargetMethod.Name)
-            && invocation.TargetMethod.ContainingType.DerivesFrom(KnownType.Org_BouncyCastle_Security_SecureRandom)
-            && state[invocation.Instance]?.HasConstraint(CryptographicSeedConstraint.Predictable) is true)
+        if ((IsSecureRandomMethod() || IsRandomGeneratorMethod())
+            && invocation.Instance is { } instance
+            && state[instance]?.HasConstraint(CryptographicSeedConstraint.Predictable) is true)
         {
             ReportIssue(invocation.WrappedOperation);
         }
-        return state;
+        return null;
+
+        bool IsSecureRandomMethod() =>
+            SecureRandomNextMethods.Contains(invocation.TargetMethod.Name)
+            && IsSecureRandom(invocation);
+
+        bool IsRandomGeneratorMethod() =>
+            invocation.TargetMethod.Name == "NextBytes"
+            && invocation.Instance is { } instance
+            && instance.Type.Is(KnownType.Org_BouncyCastle_Crypto_Prng_IRandomGenerator);
     }
+
+    private static bool IsSecureRandom(IInvocationOperationWrapper invocation) =>
+        invocation.TargetMethod.ContainingType.Is(KnownType.Org_BouncyCastle_Security_SecureRandom);
```

Comment:
This used to be `DerivesFrom`, but I changed it to `Is` because `SetSeed` and ALL the `NextXXX` methods are virtual, so the rule might not apply in a user's specific implementation.

Reply:
It makes sense. The issue appears when using the concrete types; it's not caused by the contract.